### PR TITLE
fix: Optimize SHOW FIELD KEY CARDINALITY

### PR DESCRIPTION
**Pull request by davidby-influx**
Use the `_fieldKeys` system iterator
 
Closes https://github.com/influxdata/influxdb/issues/23840

### CODE REVIEW

File: query/statement_rewriter.go [hunk 1]
```diff
@@ -69,25 +69,39 @@ func rewriteShowFieldKeyCardinalityStatement(stmt *influxql.ShowFieldKeyCardinal
 
 	return &influxql.SelectStatement{
 		Fields: []*influxql.Field{
-			{
+			&influxql.Field{
```
Comment (lesam) [inline query/statement_rewriter.go:72]: I'm surprised we need the type name here - there's a lint we can turn on somewhere for that I think.

File: query/statement_rewriter.go [hunk 2]
```diff
@@ -69,25 +69,39 @@ func rewriteShowFieldKeyCardinalityStatement(stmt *influxql.ShowFieldKeyCardinal
...
 				Alias: "count",
 			},
 		},
-		Sources:    rewriteSources2(stmt.Sources, stmt.Database),
 		Condition:  stmt.Condition,
 		Dimensions: stmt.Dimensions,
-		Offset:     stmt.Offset,
-		Limit:      stmt.Limit,
 		OmitTime:   true,
+		Sources: influxql.Sources{
```
Comment (lesam) [inline query/statement_rewriter.go:88]: This change should have required a test change for the the rewriter with some examples of the rewrite - can we add the missing tests?

**Review by lesam — CHANGES_REQUESTED** [review]

File: query/statement_rewriter.go [hunk 3]
```diff
@@ -69,25 +69,39 @@ func rewriteShowFieldKeyCardinalityStatement(stmt *influxql.ShowFieldKeyCardinal
...
+			&influxql.SubQuery{
+				Statement: &influxql.SelectStatement{
+					Fields: influxql.Fields([]*influxql.Field{
+						{Expr: &influxql.VarRef{Val: "fieldKey"}},
+						{Expr: &influxql.VarRef{Val: "fieldType"}},
+					}),
+					Sources:    rewriteSources(stmt.Sources, "_fieldKeys", stmt.Database),
+					Condition:  rewriteSourcesCondition(stmt.Sources, nil),
+					Offset:     stmt.Offset,
```
Comment (lesam) [inline query/statement_rewriter.go:97]: Feels like offset/limit should be either not accepted, or should go on the outer query?

Comment (davidby-influx) [reply query/statement_rewriter.go:97]: Outer query, I believe.  Good catch.

File: query/statement_rewriter.go (hunk 2 above)
Comment (davidby-influx) [reply query/statement_rewriter.go:88]: Do you mean just additional entries in TestRewriteStatement?

File: query/statement_rewriter.go [hunk 4]
```diff
@@ -69,25 +69,39 @@ func rewriteShowFieldKeyCardinalityStatement(stmt *influxql.ShowFieldKeyCardinal
...
-		Limit:      stmt.Limit,
 		OmitTime:   true,
+		Sources: influxql.Sources{
+			&influxql.SubQuery{
+				Statement: &influxql.SelectStatement{
+					Fields: influxql.Fields([]*influxql.Field{
+						{Expr: &influxql.VarRef{Val: "fieldKey"}},
+						{Expr: &influxql.VarRef{Val: "fieldType"}},
+					}),
+					Sources:    rewriteSources(stmt.Sources, "_fieldKeys", stmt.Database),
```
Comment (lesam) [inline query/statement_rewriter.go:95]: This source hits:

```
        case "_fieldKeys":
                return NewFieldKeysIterator(s, opt)
```

It seems we don't check for `fieldKey`, `fieldType` in the fields, we just assume those - can we also add that check in `NewFieldKeysIterator` so we don't accidentally add some subtly broken rewrite?

Comment (lesam) [reply query/statement_rewriter.go:95]: i.e. it seems wrong that the following queries have exactly the same results IIUC:

```
select foo from _fieldKeys
select fieldKey, fieldType from _fieldKeys
select fieldType, fieldKey from _fieldKeys // really wrong
```

File: query/statement_rewriter.go (hunk 2 above)
Comment (lesam) [reply query/statement_rewriter.go:88]: Yes

**Review by lesam — APPROVED** [review]
lgtm

**Pull request merged**